nncf/torch/search_blocks.py

```diff
@@ -0,0 +1,510 @@
+                potential_end_nodes.append(current_node)
+            for next_node in next_nodes:
+                q.appendleft(next_node)
+    if len(q) > 0 or len(potential_end_nodes) > 0:
+        return False
+    else:
+        for node in addit_nodes:
+            if node not in nodes:
+                return False
+        nodes.append(end_node)
+        return True
+
+def rule_not_lead_to_double_edge(sgraph: SearchGraph, start_node: SearchGraphNode, end_node: SearchGraphNode):
+    if start_node.is_dummy:
+        next_end_node = sgraph.get_next_nodes(end_node.node_key)
+        if len(next_end_node) != 0:
+            attr = sgraph._nx_graph.get_edge_data(start_node.node_key, next_end_node[0].node_key)
+        else:
+            attr = None
+    else:
+        pred_start_node = sgraph.get_prev_nodes(start_node.node_key)
+        next_end_node = sgraph.get_next_nodes(end_node.node_key)
+        if len(pred_start_node) != 0 and len(next_end_node) != 0:
+            attr = sgraph._nx_graph.get_edge_data(pred_start_node[0].node_key, next_end_node[0].node_key)
+        else:
+            attr = None
+    return True if attr is None else False
+
+def rule_not_lead_to_duplication_of_activation(sgraph: SearchGraph, start_node: SearchGraphNode, end_node: SearchGraphNode):
+    pred_start_node = sgraph.get_prev_nodes(start_node.node_key)
+    next_end_node = sgraph.get_next_nodes(end_node.node_key)
+    if len(next_end_node) == 0 or len(pred_start_node) == 0:
+        return True
+    if pred_start_node[0].is_dummy:
+        pred_start_node = sgraph.get_prev_nodes(pred_start_node[0].node_key)
+
+    if 'relu' == pred_start_node[0].node_type[-1] and 'relu' == next_end_node[0].node_type[0]:
+        return False
+    return True
+
+def comparator_for_building_block(a: BuildingBlock, b: BuildingBlock):
+    if a.end_node.bottom_id != b.end_node.bottom_id:
+        return a.end_node.bottom_id - b.end_node.bottom_id
+    else:
+        return b.start_node.main_id - a.start_node.main_id
+
+def blocks_combination_is_block(block: BuildingBlock, combination):
+
+    if block.start_node.main_id != combination[0].start_node.main_id:
+        return False
+    if block.end_node.bottom_id != combination[-1].end_node.bottom_id:
+        return False
+    i = 0
+    while i < len(combination) - 1:
+        end_i_node = combination[i].end_node
+        start_nexti_node = combination[i + 1].start_node
+        if end_i_node.node_key in start_nexti_node.node_key:
+            i += 1
+            continue
+        else:
+            return False
+    return True
+
+def search_lin_combination(block, blocks):
+    max_num = len(blocks)
+    all_combinations = []
+    for i in range(max_num, 1, -1):
+        all_combinations = list(combinations(blocks, i))
+        for combo in all_combinations:
+            if blocks_combination_is_block(block, combo):
+                return True
+    return False
+
+def remove_linear_combination(sorted_building_blocks):
```

Comment:
what do you mean by `linear combination` here? 
do you mean sequence of layer following each other?

Reply:
Yes, i do. For example `block A + block B = block C` => C- lin. combination of A and B